src/tools/wasm-ctor-eval.cpp

```diff
@@ -0,0 +1,445 @@
+            import->base == Name("STACKTOP") || // stack constants are special, we handle them
+            import->base == Name("STACK_MAX")
+          )) {
+            continue; // this is fine
+          }
+        }
+        // this global is dangerously initialized by an import, so if it is used, we must fail
+        globals.addDangerous(global->name);
+      }
+    }
+  }
+
+  enum {
+    // put the stack in some ridiculously high location
+    STACK_START = 0x40000000,
+    // use a ridiculously large stack size
+    STACK_SIZE = 32 * 1024 * 1024
+  };
+
+  std::vector<char> stack;
+
+  // create C stack space for us to use. We do *NOT* care about their contents,
+  // assuming the stack top was unwound. the memory may have been modified,
+  // but it should not be read afterwards, doing so would be undefined behavior
+  void setupEnvironment() {
+    // prepare scratch memory
+    stack.resize(STACK_SIZE);
+    // fill usable values for stack imports
+    auto total = STACK_START + STACK_SIZE;
+    globals["STACKTOP"] = Literal(int32_t(STACK_START));
+    globals["STACK_MAX"] = Literal(int32_t(STACK_START + STACK_SIZE));
+    // tell the module to accept writes up to the stack end
+    memorySize = total / Memory::kPageSize;
+  }
+
+  // flatten memory into a single segment
+  void flattenMemory() {
+    MemoryUtils::flatten(wasm.memory);
+  }
+};
+
+struct CtorEvalExternalInterface : EvallingModuleInstance::ExternalInterface {
+  Module* wasm;
+  EvallingModuleInstance* instance;
+
+  void init(Module& wasm_, EvallingModuleInstance& instance_) override {
+    wasm = &wasm_;
+    instance = &instance_;
+  }
+
+  void importGlobals(EvallingGlobalManager& globals, Module& wasm_) override {
+  }
+
+  Literal callImport(Import *import, LiteralList& arguments) override {
+    std::string extra;
+    if (import->module == "env" && import->base == "___cxa_atexit") {
+      extra = "\nrecommendation: build with -s NO_EXIT_RUNTIME=1 so that calls to atexit are not emitted";
+    }
+    throw FailToEvalException(std::string("call import: ") + import->module.str + "." + import->base.str + extra);
+  }
+
+  Literal callTable(Index index, LiteralList& arguments, WasmType result, EvallingModuleInstance& instance) override {
+    // we assume the table is not modified (hmm)
+    // look through the segments, try to find the function
+    for (auto& segment : wasm->table.segments) {
+      bool found = false;
+      Index start;
+      if (auto* c = segment.offset->dynCast<Const>()) {
+        start = c->value.getInteger();
+        found = true;
+      } else if (segment.offset->is<GetGlobal>()) {
+        // a segment loaded from a global is assumed to be loaded at 0. this is true
+        // in emscripten in the non-dynamic linking case
+        start = 0;
+        found = true;
+      }
+      if (found) {
+        auto end = start + segment.data.size();
+        if (start <= index && index < end) {
```

Comment:
Yeah, the whole `found` stuff was messy. Actually it wasn't needed. I'll remove it which simplifies a bunch, and add more comments, let me know what you think.